pkg/apb/ext_creds.go

```diff
@@ -43,54 +56,85 @@ func ExtractCredentials(podname string, ns string, runtime int) (*ExtractedCrede
+	// NOTE: kubectl exec simply sets the API path to /api when where is no
+	// Group, which is the case for pod exec.
+	clientConfig.APIPath = "/api"
 
-		if err == nil {
-			log.Notice("[%s] Bind credentials found", podname)
-			decodedOutput, err := decodeOutput(output)
+	restClient, err := rest.RESTClientFor(clientConfig)
+	if err != nil {
+		fmt.Printf("error creating rest client:")
+		fmt.Printf("%v", err.Error())
+		return nil, err
+	}
+
+	req := restClient.Post().
+		Resource("pods").
+		Name(podname).
+		Namespace(namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&v1.PodExecOptions{
+		Command: []string{GatherCredentialsCommand},
+		Stdin:   false,
+		Stdout:  true,
+		Stderr:  true,
+		TTY:     false,
+	}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(clientConfig, "POST", req.URL())
+	if err != nil {
+		fmt.Printf("error getting new remotecommand executor")
+		fmt.Printf("%v", err.Error())
+	}
+
+	for r := 1; r <= apbWatchRetries; r++ {
+		var stdoutBuffer, stderrBuffer bytes.Buffer
+		stdoutWriter := bufio.NewWriter(&stdoutBuffer)
+		stderrWriter := bufio.NewWriter(&stderrBuffer)
+
+		cmderr := exec.Stream(remotecommand.StreamOptions{
+			Stdout: stdoutWriter,
+			Stderr: stderrWriter,
+		})
+		if cmderr == nil {
+			log.Notice("[%v] bind credentials found", podname)
+			decodedOutput, err := decodeOutput(stdoutBuffer.Bytes())
 			if err != nil {
 				return nil, err
 			}
 			return buildExtractedCredentials(decodedOutput)
-		} else if podFailed {
+		}
+		//Get Pods to determine if the pod is still alive.
+		pod, err := k8scli.Client.CoreV1().Pods(namespace).Get(podname, metav1.GetOptions{})
+		if err != nil {
+			//If pod can not be found then something is very wrong.
+			log.Errorf("unable to find pod: %v in namespace: %v - err: %v", podname, namespace, err)
+			return nil, err
+		}
+		switch pod.Status.Phase {
+		case v1.PodFailed:
 			// pod has completed but is in failed state
+			log.Errorf("pod: %v in namespace: %v failed", podname, namespace)
 			return nil, fmt.Errorf("[%s] APB failed", podname)
-		} else if podCompleted && err.Error() == failedToExec.Error() {
-			log.Notice("[%s] APB completed", podname)
+		case v1.PodSucceeded:
+			log.Noticef("pod: %v in namespace: %v has been completed")
 			return nil, nil
-		} else if err.Error() == failedToExec.Error() {
-			log.Info(string(output))
-			log.Warning("[%s] Retry attempt %d: Failed to exec into the container", podname, r)
-		} else if err.Error() == credsNotAvailable.Error() {
-			log.Info(string(output))
-			log.Warning("[%s] Retry attempt %d: Bind credentials not available yet", podname, r)
-		} else {
-			log.Info(string(output))
-			log.Warning("[%s] Retry attempt %d: Failed to exec into the container", podname, r)
+		default:
+			log.Infof("command output: %v - err: %v", stdoutBuffer.String(), stderrBuffer.String())
+			log.Warningf("retry attempt: %v pod: %v in namespace: %v failed to exec into the container", r, podname, namespace)
```

Comment:
I noticed this got bumped up to a `Warning` level a while back in the old code with previous iterations of the credential extractions. I don't think this should be of a warning level since the "tried to get credentials but they weren't ready yet" branch of execution is expected to get hit a few times during a normal run of an APB.

If we agree this isn't a warning, probably drop v2 back down to `Info` as well, either in this PR or a small follow up.